extensions/standard-processors/processors/TailFile.cpp

```diff
@@ -40,60 +35,283 @@
+    openFile(file_name, offset, input_stream_, logger_);
+  }
+
+  int64_t process(std::shared_ptr<io::BaseStream> output_stream) override {
+    io::CRCStream<io::BaseStream> crc_stream{output_stream.get(), checksum_};
+
+    uint64_t num_bytes_written = 0;
+    bool found_delimiter = false;
+
+    while (hasMoreToRead() && !found_delimiter) {
+      if (begin == end) {
+        input_stream_.read(reinterpret_cast<char *>(buffer.data()), buffer.size());
+
+        std::streamsize num_bytes_read = input_stream_.gcount();
+        logger_->log_trace("Read %d bytes of input", num_bytes_read);
+
+        begin = buffer.data();
+        end = begin + num_bytes_read;
+      }
+
+      uint8_t *delimiter_pos = std::find(begin, end, input_delimiter_);
+      found_delimiter = (delimiter_pos != end);
+
+      ptrdiff_t zlen{std::distance(begin, delimiter_pos)};
+      if (found_delimiter) {
+        zlen += 1;
+      }
+      if (zlen < (std::numeric_limits<int>::min)() || zlen > (std::numeric_limits<int>::max)()) {
+        logger_->log_error("narrowing conversion failed");
+      }
+      const int len = zlen;
+
+      crc_stream.write(begin, len);
+      num_bytes_written += len;
+      begin += len;
+    }
+
+    if (found_delimiter) {
+      checksum_ = crc_stream.getCRC();
+    } else {
+      latest_flow_file_ends_with_delimiter_ = false;
+    }
+
+    return num_bytes_written;
+  }
+
+  uint64_t checksum() const {
+    return checksum_;
+  }
+
+  bool hasMoreToRead() const {
+    return begin != end || input_stream_.good();
+  }
+
+  bool useLatestFlowFile() const {
+    return latest_flow_file_ends_with_delimiter_;
+  }
+
+ private:
+  char input_delimiter_;
+  uint64_t checksum_;
+  std::ifstream input_stream_;
+  std::shared_ptr<logging::Logger> logger_;
+
+  constexpr static std::size_t BUFFER_SIZE = 4096;
+  std::array<uint8_t, BUFFER_SIZE> buffer;
+  uint8_t *begin = buffer.data();
+  uint8_t *end = buffer.data();
+
+  bool latest_flow_file_ends_with_delimiter_ = true;
+};
+
+class WholeFileReaderCallback : public OutputStreamCallback {
+ public:
+  WholeFileReaderCallback(const std::string &file_name,
+                          uint64_t offset,
+                          uint64_t checksum)
+    : checksum_(checksum),
+      logger_(logging::LoggerFactory<TailFile>::getLogger()) {
+    openFile(file_name, offset, input_stream_, logger_);
+  }
+
+  uint64_t checksum() const {
+    return checksum_;
+  }
+
+  int64_t process(std::shared_ptr<io::BaseStream> output_stream) override {
+    constexpr std::size_t BUFFER_SIZE = 4096;
+    std::array<uint8_t, BUFFER_SIZE> buffer;
+
+    io::CRCStream<io::BaseStream> crc_stream{output_stream.get(), checksum_};
+
+    uint64_t num_bytes_written = 0;
+
+    while (input_stream_.good()) {
+      input_stream_.read(reinterpret_cast<char *>(buffer.data()), buffer.size());
+
+      std::streamsize num_bytes_read = input_stream_.gcount();
+      logger_->log_trace("Read %d bytes of input", num_bytes_read);
```

Comment:
`std::streamsize` is an unspecified signed integral type used for sizes. I think we can assume that whatever value it holds can be represented by a `std::ptrdiff_t`, which has a defined format specifier.
```suggestion
        const auto num_bytes_read = input_stream_.gcount();
        logger_->log_trace("Read %td bytes of input", std::ptrdiff_t{num_bytes_read});
```

Reply:
see above